website/blog/2019-08-19-tablescan-structs.md

```diff
@@ -0,0 +1,196 @@
+This takes longer because we must read two levels of null flags, one for `export` and the other for the `license` substruct. Then we read the filing number for the positions where there is a `license` and fill in a null for the case where either `license` or the enclosing `export` struct is null.
+
+# Experiments
+
+```sql
+SELECT SUM(shipment.extendedprice), COUNT(export.license.filing_no)
+FROM exportlineitem
+WHERE shipment.partkey BETWEEN 200000 AND 400000 AND shipment.quantity < 10 AND export.s_nation IN (1, 3, 6) AND export.c_nation = 11;
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 5                   | 50                 | 1.0                     |
+| Baseline | 30                  | 309                | 5.32                    |
+
+Here we add a filter on `export`. The filters within structs are reorderable as well as the top level structs. We find that the new filter does not negatively impact running time and in fact can improve it. The best filter is evaluated first and after this all column access is sparse and only the data at interesting positions gets touched.
+
+```sql
+SELECT COUNT(*)
+FROM lineitem_map
+WHERE ints[2] BETWEEN 1000000 AND 2000000 AND ints[3] BETWEEN 100000 AND 200000;
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 35                  | 422                | 1.0                     |
+| Baseline | 50                  | 706                | 1.67                    |
+
+This case corresponds to a `lineitem` table represented as a map. This would be common if this were the feature vector for a machine learning use case, except that we would typically have hundreds of keys per map instead of the 12 here. we do the same thing as in the first query but use a direct encoded map instead. This is noticeably slower than the struct case because we must read through all the keys and values even if we do not look at them. Thus we do not save in decompression but do save in materialization and filtering. Internally, this makes a filter `ints.key IN (2, 3)`. This selects the positions in the values column that we look at. Then we make a list of filters to apply to these positions. Different positions have a different filter.  There is an extra trick in this: If there are `n` filters, e.g. 2 for 2 values that we look at out of a total of 12 values in each map and the `ith` filter is false, then we can fail the next `n - i` filters without even looking at the data.
```

Comment:
"we must read through all the keys and values even if we do not look at them" - Is there a way to explain a bit more; we do read all the keys, but we only read values that corresponding to keys 2 and 3.

Reply:
Added a commit to clarify